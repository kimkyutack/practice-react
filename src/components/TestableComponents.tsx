import React, { useState, useEffect } from "react";

// 1. 간단한 카운터 컴포넌트 (기본 테스트용)
interface CounterProps {
  initialValue?: number;
  onCountChange?: (count: number) => void;
  label?: string;
}

export const Counter: React.FC<CounterProps> = ({
  initialValue = 0,
  onCountChange,
  label = "카운터",
}) => {
  const [count, setCount] = useState(initialValue);

  useEffect(() => {
    onCountChange?.(count);
  }, [count, onCountChange]);

  const increment = () => setCount((prev) => prev + 1);
  const decrement = () => setCount((prev) => prev - 1);
  const reset = () => setCount(initialValue);

  return (
    <div data-testid="counter">
      <h3>{label}</h3>
      <p data-testid="count-display">현재 값: {count}</p>
      <div>
        <button data-testid="increment-btn" onClick={increment}>
          증가
        </button>
        <button data-testid="decrement-btn" onClick={decrement}>
          감소
        </button>
        <button data-testid="reset-btn" onClick={reset}>
          리셋
        </button>
      </div>
    </div>
  );
};

// 2. 폼 컴포넌트 (폼 테스트용)
interface FormData {
  name: string;
  email: string;
  age: string;
}

interface FormProps {
  onSubmit: (data: FormData) => void;
  initialData?: Partial<FormData>;
}

export const UserForm: React.FC<FormProps> = ({
  onSubmit,
  initialData = {},
}) => {
  const [formData, setFormData] = useState<FormData>({
    name: initialData.name || "",
    email: initialData.email || "",
    age: initialData.age || "",
  });

  const [errors, setErrors] = useState<Partial<FormData>>({});

  const validateForm = (): boolean => {
    const newErrors: Partial<FormData> = {};

    if (!formData.name.trim()) {
      newErrors.name = "이름은 필수입니다";
    }

    if (!formData.email.trim()) {
      newErrors.email = "이메일은 필수입니다";
    } else if (!/\S+@\S+\.\S+/.test(formData.email)) {
      newErrors.email = "유효한 이메일을 입력하세요";
    }

    if (!formData.age.trim()) {
      newErrors.age = "나이는 필수입니다";
    } else if (isNaN(Number(formData.age)) || Number(formData.age) < 0) {
      newErrors.age = "유효한 나이를 입력하세요";
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (validateForm()) {
      onSubmit(formData);
    }
  };

  const handleChange =
    (field: keyof FormData) => (e: React.ChangeEvent<HTMLInputElement>) => {
      setFormData((prev) => ({ ...prev, [field]: e.target.value }));
      // 에러가 있으면 실시간으로 제거
      if (errors[field]) {
        setErrors((prev) => ({ ...prev, [field]: undefined }));
      }
    };

  return (
    <form data-testid="user-form" onSubmit={handleSubmit}>
      <h3>사용자 정보 입력</h3>

      <div>
        <label htmlFor="name">이름:</label>
        <input
          id="name"
          data-testid="name-input"
          type="text"
          value={formData.name}
          onChange={handleChange("name")}
        />
        {errors.name && <span data-testid="name-error">{errors.name}</span>}
      </div>

      <div>
        <label htmlFor="email">이메일:</label>
        <input
          id="email"
          data-testid="email-input"
          type="email"
          value={formData.email}
          onChange={handleChange("email")}
        />
        {errors.email && <span data-testid="email-error">{errors.email}</span>}
      </div>

      <div>
        <label htmlFor="age">나이:</label>
        <input
          id="age"
          data-testid="age-input"
          type="number"
          value={formData.age}
          onChange={handleChange("age")}
        />
        {errors.age && <span data-testid="age-error">{errors.age}</span>}
      </div>

      <button data-testid="submit-btn" type="submit">
        제출
      </button>
    </form>
  );
};

// 3. API 호출 컴포넌트 (비동기 테스트용)
interface User {
  id: number;
  name: string;
  email: string;
}

interface UserListProps {
  onUserSelect?: (user: User) => void;
}

export const UserList: React.FC<UserListProps> = ({ onUserSelect }) => {
  const [users, setUsers] = useState<User[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const fetchUsers = async () => {
    setLoading(true);
    setError(null);

    try {
      // 실제 API 호출을 시뮬레이션
      await new Promise((resolve) => setTimeout(resolve, 1000));

      const mockUsers: User[] = [
        { id: 1, name: "김철수", email: "kim@example.com" },
        { id: 2, name: "이영희", email: "lee@example.com" },
        { id: 3, name: "박민수", email: "park@example.com" },
      ];

      setUsers(mockUsers);
    } catch (err) {
      setError("사용자 목록을 불러오는데 실패했습니다");
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchUsers();
  }, []);

  if (loading) {
    return <div data-testid="loading">로딩 중...</div>;
  }

  if (error) {
    return (
      <div data-testid="error">
        <p>{error}</p>
        <button data-testid="retry-btn" onClick={fetchUsers}>
          다시 시도
        </button>
      </div>
    );
  }

  return (
    <div data-testid="user-list">
      <h3>사용자 목록</h3>
      {users.length === 0 ? (
        <p data-testid="no-users">사용자가 없습니다</p>
      ) : (
        <ul>
          {users.map((user) => (
            <li
              key={user.id}
              data-testid={`user-item-${user.id}`}
              onClick={() => onUserSelect?.(user)}
              style={{ cursor: "pointer", padding: "5px" }}
            >
              {user.name} ({user.email})
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

// 4. 조건부 렌더링 컴포넌트 (조건부 테스트용)
interface ConditionalComponentProps {
  isLoggedIn: boolean;
  userRole?: "admin" | "user" | "guest";
  showDetails: boolean;
}

export const ConditionalComponent: React.FC<ConditionalComponentProps> = ({
  isLoggedIn,
  userRole = "guest",
  showDetails,
}) => {
  return (
    <div data-testid="conditional-component">
      {isLoggedIn ? (
        <div data-testid="logged-in-content">
          <h3>환영합니다!</h3>
          {userRole === "admin" && (
            <div data-testid="admin-panel">
              <h4>관리자 패널</h4>
              <p>관리자만 볼 수 있는 내용입니다</p>
            </div>
          )}
          {showDetails && (
            <div data-testid="user-details">
              <p>사용자 역할: {userRole}</p>
              <p>상세 정보가 표시됩니다</p>
            </div>
          )}
        </div>
      ) : (
        <div data-testid="login-prompt">
          <h3>로그인이 필요합니다</h3>
          <p>계속하려면 로그인해주세요</p>
        </div>
      )}
    </div>
  );
};
