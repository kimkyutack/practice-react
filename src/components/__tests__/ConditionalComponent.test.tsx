import React from "react";
import { render, screen } from "@testing-library/react";
import { ConditionalComponent } from "../TestableComponents";

// ConditionalComponent 테스트
describe("ConditionalComponent", () => {
  // 1. 로그인하지 않은 상태 테스트
  test("로그인하지 않은 상태에서 로그인 프롬프트가 표시된다", () => {
    render(<ConditionalComponent isLoggedIn={false} showDetails={false} />);

    expect(screen.getByTestId("conditional-component")).toBeInTheDocument();
    expect(screen.getByTestId("login-prompt")).toBeInTheDocument();
    expect(screen.getByText("로그인이 필요합니다")).toBeInTheDocument();
    expect(screen.getByText("계속하려면 로그인해주세요")).toBeInTheDocument();

    // 로그인된 콘텐츠는 표시되지 않아야 함
    expect(screen.queryByTestId("logged-in-content")).not.toBeInTheDocument();
  });

  // 2. 로그인된 상태 테스트
  test("로그인된 상태에서 환영 메시지가 표시된다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="user"
        showDetails={false}
      />
    );

    expect(screen.getByTestId("conditional-component")).toBeInTheDocument();
    expect(screen.getByTestId("logged-in-content")).toBeInTheDocument();
    expect(screen.getByText("환영합니다!")).toBeInTheDocument();

    // 로그인 프롬프트는 표시되지 않아야 함
    expect(screen.queryByTestId("login-prompt")).not.toBeInTheDocument();
  });

  // 3. 관리자 패널 테스트
  test("관리자 역할일 때 관리자 패널이 표시된다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={false}
      />
    );

    expect(screen.getByTestId("admin-panel")).toBeInTheDocument();
    expect(screen.getByText("관리자 패널")).toBeInTheDocument();
    expect(
      screen.getByText("관리자만 볼 수 있는 내용입니다")
    ).toBeInTheDocument();
  });

  test("일반 사용자 역할일 때 관리자 패널이 표시되지 않는다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="user"
        showDetails={false}
      />
    );

    expect(screen.queryByTestId("admin-panel")).not.toBeInTheDocument();
  });

  test("게스트 역할일 때 관리자 패널이 표시되지 않는다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="guest"
        showDetails={false}
      />
    );

    expect(screen.queryByTestId("admin-panel")).not.toBeInTheDocument();
  });

  // 4. 상세 정보 표시 테스트
  test("showDetails가 true일 때 상세 정보가 표시된다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="user"
        showDetails={true}
      />
    );

    expect(screen.getByTestId("user-details")).toBeInTheDocument();
    expect(screen.getByText("사용자 역할: user")).toBeInTheDocument();
    expect(screen.getByText("상세 정보가 표시됩니다")).toBeInTheDocument();
  });

  test("showDetails가 false일 때 상세 정보가 표시되지 않는다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="user"
        showDetails={false}
      />
    );

    expect(screen.queryByTestId("user-details")).not.toBeInTheDocument();
  });

  // 5. 복합 조건 테스트
  test("관리자이면서 상세 정보가 표시되는 경우", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={true}
      />
    );

    // 로그인된 콘텐츠
    expect(screen.getByTestId("logged-in-content")).toBeInTheDocument();
    expect(screen.getByText("환영합니다!")).toBeInTheDocument();

    // 관리자 패널
    expect(screen.getByTestId("admin-panel")).toBeInTheDocument();
    expect(screen.getByText("관리자 패널")).toBeInTheDocument();

    // 상세 정보
    expect(screen.getByTestId("user-details")).toBeInTheDocument();
    expect(screen.getByText("사용자 역할: admin")).toBeInTheDocument();
  });

  // 6. 기본값 테스트
  test("userRole이 제공되지 않으면 guest로 기본값이 설정된다", () => {
    render(<ConditionalComponent isLoggedIn={true} showDetails={true} />);

    expect(screen.getByText("사용자 역할: guest")).toBeInTheDocument();
    expect(screen.queryByTestId("admin-panel")).not.toBeInTheDocument();
  });

  // 7. 접근성 테스트
  test("모든 텍스트가 스크린 리더에서 접근 가능하다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={true}
      />
    );

    // 모든 중요한 텍스트가 존재하는지 확인
    expect(screen.getByText("환영합니다!")).toBeInTheDocument();
    expect(screen.getByText("관리자 패널")).toBeInTheDocument();
    expect(
      screen.getByText("관리자만 볼 수 있는 내용입니다")
    ).toBeInTheDocument();
    expect(screen.getByText("사용자 역할: admin")).toBeInTheDocument();
    expect(screen.getByText("상세 정보가 표시됩니다")).toBeInTheDocument();
  });

  // 8. 스냅샷 테스트
  test("로그인하지 않은 상태의 스냅샷이 일치한다", () => {
    const { container } = render(
      <ConditionalComponent isLoggedIn={false} showDetails={false} />
    );

    expect(container).toMatchSnapshot();
  });

  test("관리자 로그인 상태의 스냅샷이 일치한다", () => {
    const { container } = render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={true}
      />
    );

    expect(container).toMatchSnapshot();
  });

  // 9. 엣지 케이스 테스트
  test("빈 문자열 userRole이 제공되면 guest로 처리된다", () => {
    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole={"" as any}
        showDetails={true}
      />
    );

    expect(screen.getByText("사용자 역할: ")).toBeInTheDocument();
    expect(screen.queryByTestId("admin-panel")).not.toBeInTheDocument();
  });

  // 10. 성능 테스트
  test("조건부 렌더링이 빠르게 작동한다", () => {
    const startTime = performance.now();

    render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={true}
      />
    );

    const endTime = performance.now();
    const renderTime = endTime - startTime;

    // 100ms 이내에 렌더링되어야 함
    expect(renderTime).toBeLessThan(100);
  });

  // 11. 메모리 사용량 테스트
  test("조건부 렌더링이 메모리 효율적으로 작동한다", () => {
    const initialMemory = (performance as any).memory?.usedJSHeapSize || 0;

    const { unmount } = render(
      <ConditionalComponent
        isLoggedIn={true}
        userRole="admin"
        showDetails={true}
      />
    );

    unmount();

    const finalMemory = (performance as any).memory?.usedJSHeapSize || 0;
    const memoryIncrease = finalMemory - initialMemory;

    // 메모리 증가가 합리적인 범위 내에 있어야 함 (1MB 이하)
    expect(memoryIncrease).toBeLessThan(1024 * 1024);
  });
});
