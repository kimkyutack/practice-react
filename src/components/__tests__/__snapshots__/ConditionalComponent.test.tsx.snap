// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalComponent 관리자 로그인 상태의 스냅샷이 일치한다 1`] = `
<div>
  <div
    data-testid="conditional-component"
  >
    <div
      data-testid="logged-in-content"
    >
      <h3>
        환영합니다!
      </h3>
      <div
        data-testid="admin-panel"
      >
        <h4>
          관리자 패널
        </h4>
        <p>
          관리자만 볼 수 있는 내용입니다
        </p>
      </div>
      <div
        data-testid="user-details"
      >
        <p>
          사용자 역할: 
          admin
        </p>
        <p>
          상세 정보가 표시됩니다
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConditionalComponent 로그인하지 않은 상태의 스냅샷이 일치한다 1`] = `
<div>
  <div
    data-testid="conditional-component"
  >
    <div
      data-testid="login-prompt"
    >
      <h3>
        로그인이 필요합니다
      </h3>
      <p>
        계속하려면 로그인해주세요
      </p>
    </div>
  </div>
</div>
`;
